File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.java
import java.util.*;

public class Solution {

    private static final Map<String, Integer> mineralIdxTable = Map.of(
            "diamond", 0,
            "iron", 1,
            "stone", 2
    );
    private static final int[][] costTable = {
            {1, 1, 1},
            {5, 1, 1},
            {25, 5, 1}
    };

    public static int answer = Integer.MAX_VALUE;

    public int solution(int[] picks, String[] minerals) {
        int totalPicks = 0;
        for (int i = 0; i < picks.length; i++) {
            totalPicks += picks[i];
        }
        miningMinerals(0, 0, 0, 0, totalPicks, picks, minerals);
        return answer;
    }

    private void miningMinerals(int pickIdx, int mineralIdx, int fatigue, int count, int totalPicks, int[] picks, String[] minerals) {

        if (mineralIdx >= minerals.length || (count == 0 && totalPicks == 0)) {
            answer = Math.min(answer, fatigue);
            return;
        }

        int mineral = mineralIdxTable.get(minerals[mineralIdx]);

        if (mineralIdx % 5 != 0) {
            int cost = costTable[pickIdx][mineral];
            miningMinerals(pickIdx, mineralIdx + 1, fatigue + cost, count - 1, totalPicks, picks, minerals);
            return;
        } else {
            for (int i = 0; i < picks.length; i++) {
                if (picks[i] == 0) continue;

                picks[i]--;
                int cost = costTable[i][mineral];
                miningMinerals(i, mineralIdx + 1, fatigue + cost, 4, totalPicks - 1, picks, minerals);
                picks[i]++;
            }
        }
    }
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.cpp
#include <string>
#include <vector>
#include <map>
#include <cmath>

using namespace std;
map<string, int> m = { {"diamond", 0}, {"iron", 1},{"stone", 2} };
int answer = 1000000;

void Work(vector<int> picks, vector<string> minerals, int depth, int tired, int count, int currentPick, int total)
{
    if (depth >= minerals.size() || (count == 0 && total == 0))
    {
        answer = min(tired, answer);
        return;
    }

    string currentMineral = minerals[depth];

    if (count > 0)
    {
        int t = pow(5, max(currentPick - m[currentMineral], 0));
        Work(picks, minerals, depth + 1, tired + t, count - 1, currentPick, total);
    }
    else
    {
        for (int i = 0; i < picks.size(); ++i)
        {
            if (picks[i] <= 0)
                continue;

            currentPick = i;
            --picks[i];
            count = 5;
            int t = pow(5, max(currentPick - m[currentMineral], 0));
            Work(picks, minerals, depth + 1, tired + t, count - 1, currentPick, total - 1);
            ++picks[i];
        }
    }
}

int solution(vector<int> picks, vector<string> minerals) {

    int total = 0;
    
    for(int i = 0; i < picks.size(); ++i)
    {
        total += picks[i];
    }
    
    Work(picks, minerals, 0, 0, 0, 0, total);

    return answer;
}